Add salt 3000.2 recipe

shruthi-ravi opened the pull request:
RTOS would like to upgrade the salt packages supported on NILRT to match the SystemLink host. This patchset upgrades the salt recipe to v3000.2 and is based on the salt v2018.3 recipe in sumo.

@andzn has tested that the resulting salt 3000 packages work as expected.
This patchset does not include the salt-ptest. There are issues getting those tests to run in hardknott that I am working through at this time.

AzDO [1571913](https://ni.visualstudio.com/DevCentral/_workitems/edit/1571913).

@ni/rtos

amstewart committed:
salt: add transconf subpackage

Add a subpackage to install transconf hooks for migrating salt
configuration.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>
Acked-by: Bill Pittman <bill.pittman@ni.com>
Natinst-ReviewBoard-ID: 305120

Rebased-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: Add back salt 2018.3 recipe and files

Add back the salt 2018.3 recipe as a salt 3000.1 recipe, and
associated files. The actual salt 3000.1 recipe will be developed
with this as the base.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: Update recipe to point to ni/salt 3000.2

Update SRC_URI and CHKSUM corresponding to the
ni/salt:ni/master/3000.2 branch and set to AUTOREV.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: Replace dependency on lsb with lsb-release

lsb is deprecated and is no longer shipped. So depend on lsb-release
instead.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: Refresh python location hashbang patch

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: Set `S` to git work directory

The existing python hashbang patch is relative to the git directory.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: Use version-agnostic variable to reference python dir

The files in the salt Python 3.9 lib directory were not being
packaged correctly. Without this change, no package owned the files
installed into the salt Python 3.9 directory causing a bitbake
installed-vs-shipped QA error.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: Fix Python import error for the distro package

Add missing python3-distro as a build time dependency. This
dependency is required in Python 3.8 and above.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: Miscellaneous additions from the 2018.3 bbappends

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: Enable only the tcp feature.

This change follows the salt 2018.3 bbappends file.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

shruthi-ravi committed:
salt: add logrotate conf file

Add a custom logrotate file to more aggressively limit the log size.
This change is from the salt 2018.3 bbappend recipe.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

amstewart commented on recipes-support/salt/salt_3000.2.bb:
```diff
@@ -0,0 +1,203 @@
+
+SRCREV = "${AUTOREV}"
+
+S="${WORKDIR}/git"
```
NIT: spaces before and after the `=`.

amstewart commented on recipes-support/salt/salt_3000.2.bb:
```diff
@@ -0,0 +1,203 @@
+SECTION = "admin"
+LICENSE = "Apache-2.0"
+LIC_FILES_CHKSUM = "file://LICENSE;md5=c996f5a78d858a52c894fa3f4bec68c1"
+DEPENDS = "\
+           python3-msgpack \
+           python3-pyyaml \
+           python3-jinja2 \
+           python3-markupsafe \
+"
+
+DEPENDS = "\
+           python3-distro-native \
+"
```
Is there a reason that these are split? And that you're overwriting the depends instead of doing `DEPENDS_${PN} += ...`?

shruthi-ravi replied:
I wanted to keep the native dependency separate, but did not realize I was overwriting the variable. Good catch, thanks.

amstewart commented on recipes-support/salt/salt_3000.2.bb:
```diff
@@ -0,0 +1,203 @@
+SECTION = "admin"
+LICENSE = "Apache-2.0"
+LIC_FILES_CHKSUM = "file://LICENSE;md5=c996f5a78d858a52c894fa3f4bec68c1"
+DEPENDS = "\
```
NIT: While we're authoring this file, could you update the tab indent-styling. We're converging on the hanging-tab style:
````
VARIABLE = "\
<tab>foobar \
"
```

shruthi-ravi replied:
Can you clarify? Do you want a single tab instead of multiple? Do you want to replace spaces with a tab?

amstewart replied:
Both or either; depending on which variable assignment we're talking about.

shruthi-ravi replied:
Ok. That did not do much for my confusion :). I took a pass at it. Take a look and lmk if you would like additional changes.

amstewart replied:
You got it. Looks good.

amstewart commented on recipes-support/salt/salt_3000.2.bb:
```diff
@@ -55,6 +55,37 @@ PACKAGES += "\
+    python3-psutil \
+"
+
+# these dependencies should NOT be added to the salt bb recipe as they're added
+# here in the bbappend for the NIFeeds ni-skyline packages. In the future these
+# will be removed and skyline packages made to depend on them directly.
```
Do the SL folks have any more context on this effort? Do we still need to add these?

shruthi-ravi replied:
Yes, still needed. I asked them and they hadn't made the required changes to the skyline packages.

amstewart commented on recipes-support/salt/files/set_python_location_hashbang.patch:
```diff
@@ -0,0 +1,107 @@
+Upstream-Status: Pending
+
+# The Salt SysV scripts require that the process name of the salt
+# components have the form "salt-<component>".
+# The current python shebangs on the salt components scripts spwans
+# processes that are generically named python. Changed shebang so
+# process names will be identifiable by the init scripts.
```
Since we own the sysVinit scripts, can we just fixup the initscript's logic to correctly handle the salt python scripts? The shebang lines they're using is best-practice, and I would prefer a local modification to our initscript over applying a `.patch` to a dozen files.

shruthi-ravi replied:
I'm not sure how we can match a python process in an init script, if not for the process name.
Also, I think a change like this would generate more churn than necessary at this time.

amstewart replied:
I'm pretty sure `start-stop-daemon` could still match the process based on the `PIDFILE` alone, and the `--name` argument is just there to catch the case that the original process was killed and another process happened to startup with the same PID. That being said, I understand the complication with running a python script which uses `env` as an indirection layer and it seeting `/proc/$pid/comm` to `python` instead of the script name.

I still think there is a better way to handle this, but after searching it isn't obvious and I'm not willing to block these changes on it.

amstewart reviewed (changes requested):
```
salt: Use version-agnostic variable to reference python dir

The files in the current Python 3.9 lib directory were not being
packaged correctly. This change fixes the build breakage.
```

Could you add some detail to this commit message about how it was being incorrectly packaged.

andzn commented on recipes-support/salt/salt_3000.2.bb:
```diff
@@ -0,0 +1,203 @@
+
+# Note ${PN}-tests must be before ${PN}-common in the PACKAGES variable
+# in order for ${PN}-tests to own the correct FILES.
+PACKAGES += "\
```
FYI: out of all those packages we only use (and test) salt-minion and salt-common, so if it helps you could get rid of the others.

shruthi-ravi replied:
Ok, I'll see how well I can clean up the remaining packages.

shruthi-ravi replied:
I've looked into this some more, and I won't be able to remove the other packages cleanly. The files - salt-syndic, salt-cloud, etc.- currently owned by the other packages are installed by distutils. And if there are no corresponding salt packages in the recipe that claim ownership of these files, bitbake errors out during the build. So I think I will let the remaining packages be as is, instead of trying to remove these files from the installed directories.
@amstewart  thoughts?

amstewart replied:
It will probably be more maintenance burden to remove the packages than to maintain them - since bitbake will normally complain if all of the files installed to the sysroot don't get packaged *somewhere*. So I would recommend you leave them in the recipe.

shruthi-ravi committed:
salt: Fix indentation.

Signed-off-by: Shruthi Ravichandran <shruthi.ravichandran@ni.com>

amstewart commented:
@shruthi-ravi Could you post some patch-rev notes with a summary of what changed in the force-push?

shruthi-ravi commented:
Patch v2

- Added more context to the `salt: Use version-agnostic variable to reference python dir` commit
- Fixed variable assignment spacing and variable indentation in the salt recipe
- Fixed bug where I was overwriting the DEPENDS with the native python-distro dependency
